pkg/sql/sessionprotectedts/session_protected_ts_test.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptreconcile"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptstorage"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessionprotectedts"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func createSessionAndProtect(
+	ctx context.Context,
+	t *testing.T,
+	s0 serverutils.ApplicationLayerInterface,
+	sqlDB *sqlutils.SQLRunner,
+	pts protectedts.Storage,
+) (*gosql.DB, *ptpb.Record) {
+	t.Helper()
+
+	getUserConn := func(t *testing.T, server serverutils.ApplicationLayerInterface) *gosql.DB {
+		pgURL := url.URL{
+			Scheme: "postgres",
+			User:   url.UserPassword("foo", "hunter2"),
+			Host:   server.AdvSQLAddr(),
+		}
+		db, err := gosql.Open("postgres", pgURL.String())
+		require.NoError(t, err)
+		return db
+	}
+
+	// Start the first SQL session that will lay down a PTS record.
+	targetDB := getUserConn(t, s0)
+	sqlutils.MakeSQLRunner(targetDB).Exec(t, `SELECT 1`)
+
+	// Fetch the IDLE session's ID.
+	var idleSessionID string
+	sqlDB.QueryRow(t, `SELECT session_id FROM crdb_internal.cluster_sessions WHERE last_active_query LIKE 'SELECT 1%' ORDER BY session_start DESC LIMIT 1`).Scan(&idleSessionID)
+	recordID := uuid.MakeV4()
+	rec := sessionprotectedts.MakeRecord(recordID, []byte(idleSessionID), s0.Clock().Now(), ptpb.MakeClusterTarget())
+	require.NoError(t, pts.Protect(ctx, rec))
+
+	return targetDB, rec
+}
+
+func TestSessionProtectedTimestampReconciler(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	defer log.Scope(t).Close(t)
+
+	testCluster := serverutils.StartCluster(t, 1, base.TestClusterArgs{})
+	defer testCluster.Stopper().Stop(ctx)
+	s0 := testCluster.Server(0).ApplicationLayer()
+
+	sqlDB := sqlutils.MakeSQLRunner(s0.SQLConn(t, "defaultdb"))
+	_ = sqlDB.Exec(t, `
+CREATE USER foo with password 'hunter2';
+`)
+
+	insqlDB := s0.InternalDB().(isql.DB)
+	ptp := s0.ExecutorConfig().(sql.ExecutorConfig).ProtectedTimestampProvider
+	pts := ptstorage.WithDatabase(ptp, insqlDB)
+
+	settings := cluster.MakeTestingClusterSettings()
+	r := ptreconcile.New(settings, insqlDB, ptp,
+		ptreconcile.StatusFuncs{
+			sessionprotectedts.SessionMetaType: sessionprotectedts.MakeStatusFunc(),
+		})
+	require.NoError(t, r.StartReconciler(ctx, s0.AppStopper()))
+
+	// Create two sessions, these will be IDLE as they are running SELECT queries.
```

Comment:
Is the `as they are` clause correct here. Aren't they idle because they _aren't_ actively running a query?

Reply:
yup needed to be fixed, done